File: src/main/java/dev/cele/asa_sm/ui/components/ServerTab.java
package dev.cele.asa_sm.ui.components;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.formdev.flatlaf.FlatClientProperties;
import dev.cele.asa_sm.Const;
import dev.cele.asa_sm.config.SpringApplicationContext;
import dev.cele.asa_sm.dto.AsaServerConfigDto;
import dev.cele.asa_sm.services.CommandRunnerService;
import dev.cele.asa_sm.services.IniSerializerService;
import dev.cele.asa_sm.services.SteamCMDService;
import dev.cele.asa_sm.ui.components.forms.SliderWithText;
import dev.cele.asa_sm.ui.components.server_tab_accordions.*;
import dev.cele.asa_sm.ui.frames.ProcessDialog;
import dev.cele.asa_sm.ui.listeners.SimpleDocumentListener;
import lombok.Getter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.core.env.Environment;

import javax.swing.*;
import javax.swing.text.JTextComponent;
import java.awt.*;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.function.Consumer;

import static java.util.stream.Collectors.toMap;

public class ServerTab extends JPanel {

    //region Autowired stuff from spring
    private final SteamCMDService steamCMDService = SpringApplicationContext.autoWire(SteamCMDService.class);
    private final CommandRunnerService commandRunnerService = SpringApplicationContext.autoWire(CommandRunnerService.class);
    private final ObjectMapper objectMapper = SpringApplicationContext.autoWire(ObjectMapper.class);
    private final IniSerializerService iniSerializerService = SpringApplicationContext.autoWire(IniSerializerService.class);
    private Logger log = LoggerFactory.getLogger(ServerTab.class);

    private final Environment environment = SpringApplicationContext.autoWire(Environment.class);
    private final boolean isDev = Arrays.asList(environment.getActiveProfiles()).contains("dev");

    //endregion

    //region UI components
    private final JPanel scrollPaneContent;
    private final TopPanel topPanel;

    //endregion


    private Thread serverThread = null;
    @Getter
    private AsaServerConfigDto configDto;

    public ServerTab(AsaServerConfigDto configDto) {
        this.configDto = configDto;

        //region setup UI components
        //region initial setup, var scrollPaneContent = JPanel()
        setLayout(new BorderLayout());

        GridBagConstraints globalVerticalGBC = new GridBagConstraints();
        globalVerticalGBC.fill = GridBagConstraints.HORIZONTAL;
        globalVerticalGBC.anchor = GridBagConstraints.BASELINE;
        globalVerticalGBC.gridwidth = GridBagConstraints.REMAINDER;
        globalVerticalGBC.weightx = 1.0;
        globalVerticalGBC.insets = new Insets(5, 5, 5, 0);

        //create a JScrollPane and a content panel
        scrollPaneContent = new JPanel();
        scrollPaneContent.setLayout(new GridBagLayout());
        JScrollPane scrollPane = new JScrollPane(scrollPaneContent);
        scrollPane.getVerticalScrollBar().setUnitIncrement(10);
        scrollPane.setVerticalScrollBarPolicy(JScrollPane.VERTICAL_SCROLLBAR_ALWAYS);
        scrollPane.setHorizontalScrollBarPolicy(JScrollPane.HORIZONTAL_SCROLLBAR_NEVER);
        add(scrollPane, BorderLayout.CENTER);
        //endregion

        //top panel
        topPanel = new TopPanel(configDto);
        scrollPaneContent.add(topPanel.$$$getRootComponent$$$(), globalVerticalGBC);

        //create a group named "Administration"
        var administrationAccordion = new AdministrationAccordion(configDto);
        scrollPaneContent.add(administrationAccordion.$$$getRootComponent$$$(), globalVerticalGBC);

        //... other accordion groups ...
        if(isDev){
            var rulesAccordion = new RulesAccordion(configDto);
            scrollPaneContent.add(rulesAccordion.$$$getRootComponent$$$(), globalVerticalGBC);
        }

        var chatAndNotificationsAccordion = new ChatAndNotificationsAccordion(configDto);
        scrollPaneContent.add(chatAndNotificationsAccordion.$$$getRootComponent$$$(), globalVerticalGBC);

        var hudAndVisualsAccordion = new HUDAndVisuals(configDto);
        scrollPaneContent.add(hudAndVisualsAccordion.$$$getRootComponent$$$(), globalVerticalGBC);

        if(isDev){
            var playerSettingsAccordion = new PlayerSettingsAccordion(configDto);
            scrollPaneContent.add(playerSettingsAccordion.$$$getRootComponent$$$(), globalVerticalGBC);
        }

        //create an empty filler panel that will fill the remaining space if there's any
        JPanel fillerPanel = new JPanel();
        fillerPanel.setPreferredSize(new Dimension(0, 0));
        scrollPaneContent.add(fillerPanel, gbcClone(globalVerticalGBC, gbc -> gbc.weighty = 1.0));
        //endregion


        if(detectInstalled()){
            //if the server is already installed read the ini files and populate the configDto,
            //the ini files have priority so if you create a profile for an already existing server
            //the interface will be populated with the values from your existing server
            readAllIniFiles();
            if(configDto.getJustImported()){

                //import extra settings from INI Files that are both on INI files and on the configDto
                configDto.setServerPassword(configDto.getGameUserSettingsINI().getServerSettings().getServerPassword());
                configDto.setServerAdminPassword(configDto.getGameUserSettingsINI().getServerSettings().getServerAdminPassword());
                configDto.setServerSpectatorPassword(configDto.getGameUserSettingsINI().getServerSettings().getSpectatorPassword());

                configDto.setServerName(configDto.getGameUserSettingsINI().getSessionSettings().getSessionName());

                configDto.setModIds(configDto.getGameUserSettingsINI().getServerSettings().getActiveMods());

                configDto.setRconEnabled(configDto.getGameUserSettingsINI().getServerSettings().getRconEnabled());
                configDto.setRconPort(configDto.getGameUserSettingsINI().getServerSettings().getRconPort());
                configDto.setRconServerLogBuffer(configDto.getGameUserSettingsINI().getServerSettings().getRconServerGameLogBuffer());
            }
        }


        configDto.addUnsavedChangeListener((unsaved) -> {
            //get tabbed pane and set title with a star if unsaved
            JTabbedPane tabbedPane = (JTabbedPane) SwingUtilities.getAncestorOfClass(JTabbedPane.class, this);
            if(tabbedPane != null){
                int index = tabbedPane.indexOfComponent(this);
                if(index != -1){
                    tabbedPane.setTitleAt(index, configDto.getProfileName() + (unsaved ? " *" : ""));
                }
            }

            //set the save button outline to red if unsaved
            if(unsaved){
                topPanel.saveButton.putClientProperty(FlatClientProperties.OUTLINE, FlatClientProperties.OUTLINE_ERROR);
            } else {
                topPanel.saveButton.putClientProperty(FlatClientProperties.OUTLINE, null);
            }
        });

        SwingUtilities.invokeLater(() -> {
            setupListenersForUnsaved(this);
            setupAccordionsExpansion();
        });
    }

    void setupAccordionsExpansion(){
        var accordions = new ArrayList<AccordionTopBar>();

        //loop over all the children of the container
        for (Component component : scrollPaneContent.getComponents()) {
            //if the component is a container, has a borderlayout, and his top component is a AccordionTopBar
            if(component instanceof Container &&
                    ((Container) component).getLayout() instanceof BorderLayout
            ){
                var topComponent = ((BorderLayout) ((Container) component).getLayout()).getLayoutComponent(BorderLayout.NORTH);
                if(topComponent instanceof AccordionTopBar){
                    accordions.add((AccordionTopBar) topComponent);
                }
            }
        }

        new Thread(() -> {
            try {
                Thread.sleep(1500);
                SwingUtilities.invokeLater(() -> {
                    for (var accordion : accordions) {
                        //TODO: use the text from the accordion to check if it's expanded or not and save it to a json or something?
                        accordion.setExpanded(false);
                    }
                });
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        }).start();
    }

    void setupListenersForUnsaved(Container container){
        //loop over all the children of the container
        for (Component component : container.getComponents()) {
            //if the component is a container, call this function recursively

            if(component instanceof SliderWithText){
                ((SliderWithText) component).addChangeListener(e -> configDto.setUnsaved(true));
            } else if (component instanceof JTextComponent){
                ((JTextComponent) component).getDocument().addDocumentListener(new SimpleDocumentListener(text -> configDto.setUnsaved(true)));
            } else if (component instanceof JCheckBox){
                ((JCheckBox) component).addActionListener(e -> configDto.setUnsaved(true));
            } else if (component instanceof JComboBox){
                ((JComboBox<?>) component).addActionListener(e -> configDto.setUnsaved(true));
            } else if (component instanceof JSpinner){
                ((JSpinner) component).addChangeListener(e -> configDto.setUnsaved(true));
            }else if(component instanceof Container){
                setupListenersForUnsaved((Container) component);
            }

        }
    }

    private void readAllIniFiles(){
        var gameUserSettingsIniFile = configDto.getServerPath()
                .resolve("ShooterGame")
                .resolve("Saved")
                .resolve("Config")
                .resolve("WindowsServer")
                .resolve("GameUserSettings.ini")
                .toFile();

        if(gameUserSettingsIniFile.exists()){
            iniSerializerService.readIniFile(configDto.getGameUserSettingsINI(), gameUserSettingsIniFile);
        }

        //TODO: read Game.ini
    }

    private void writeAllIniFiles(){
        var gameUserSettingsIniFile = configDto.getServerPath()
                .resolve("ShooterGame")
                .resolve("Saved")
                .resolve("Config")
                .resolve("WindowsServer")
                .resolve("GameUserSettings.ini")
                .toFile();

        iniSerializerService.writeIniFile(configDto.getGameUserSettingsINI(), gameUserSettingsIniFile);
    }


    private boolean detectInstalled(){
        //check if the server is installed
        if (Files.exists(configDto.getServerPath())) {
            topPanel.installVerifyButton.setText("Verify/Update");
            topPanel.openInstallLocationButton.setEnabled(true);
            topPanel.installedLocationLabel.setText(configDto.getServerPath().toString());
            topPanel.startButton.setEnabled(true);

            readVersionNumber();

            return true;
        } else {
            topPanel.installVerifyButton.setText("Install");
            topPanel.openInstallLocationButton.setEnabled(false);
            topPanel.installedLocationLabel.setText("Not installed yet");
            topPanel.startButton.setEnabled(false);
            return false;
        }
    }

    private void readVersionNumber() {
        try {
            //read version from log file (ShooterGame\Saved\Logs\ShooterGame.log)
            var logFile = configDto.getServerPath()
                    .resolve("ShooterGame")
                    .resolve("Saved")
                    .resolve("Logs")
                    .resolve("ShooterGame.log")
                    .toFile();

            //read file line by line, not all at once, because the file can be huge
            var bufferedReader = new BufferedReader(new FileReader(logFile));
            String line;

            var ARK_LOG_VERSION_DIVIDER = "ARK Version: ";

            while((line = bufferedReader.readLine()) != null){
                if(line.contains(ARK_LOG_VERSION_DIVIDER)){
                    var versionParts = line.split(ARK_LOG_VERSION_DIVIDER);
                    var version = "?";
                    if(versionParts.length == 2){
                        version = versionParts[1];
                    }

                    log.info("Detected version: " + version);

                    topPanel.installedVersionLabel.setText(version);
                    break;
                }
            }
            bufferedReader.close();
        }catch (Exception e){
            log.error("Error reading version from log file", e);
        }
    }

    public void install(){
        var wasInstalled = detectInstalled();
        ProcessDialog processDialog = new ProcessDialog(
                (JFrame) SwingUtilities.getAncestorOfClass(JFrame.class, this),
                wasInstalled ? "Verifying/Updating server..." : "Installing server...",
                steamCMDService.downloadVerifyServerCommand(configDto.getServerPath()),
                result -> {
                    if(!wasInstalled){
                        configDto.setUnsaved(true);
                    }
                    detectInstalled();
                }
        );
        processDialog.setVisible(true);
    }


    public boolean isServerRunning(){
        return serverThread != null && serverThread.isAlive();
    }

    public void save() {
        //save the configDto to a json file
        Path configFile = Path.of("data" + File.separator + "profiles" + File.separator + configDto.getGuid() + ".json");
        try {
            objectMapper.writerWithDefaultPrettyPrinter().writeValue(configFile.toFile(), configDto);
            log.info("Saved configDto to " + configFile + " file");
            configDto.setUnsaved(false);
            writeAllIniFiles();
        } catch (Exception e) {
            log.error("Error saving configDto to file", e);
            JOptionPane.showMessageDialog(this, "Error saving configDto to file", "Error", JOptionPane.ERROR_MESSAGE);
        }
    }

    public void startServer() {
        //check if the server is running
        if (isServerRunning()) {
            serverThread.interrupt();
            topPanel.startButton.setText("Start");
            return;
        }

        if(getConfigDto().getJustImported()){
            var result = JOptionPane.showConfirmDialog(this, """
            You have just imported this server, starting it with wrong settings may damage data
            (for example check mods ids and similar stuff)
            Are you sure you checked all and you want to start it?"""
                    , "Warning",
                    JOptionPane.YES_NO_OPTION,
                    JOptionPane.WARNING_MESSAGE
            );

            configDto.setJustImported(false);
            if(result != JOptionPane.YES_OPTION){
                save();
            }
        }

        //start the server
        serverThread = new Thread(() -> {
            commandRunnerService.runCommand(configDto.getCommand());
            SwingUtilities.invokeLater(() ->
                    topPanel.startButton.setText("Start")
            );
        });

        //update the version number
        new Thread(() -> {
            try {
                Thread.sleep(1000);
                SwingUtilities.invokeLater(() -> readVersionNumber());
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        }).start();

        serverThread.start();
        topPanel.startButton.setText("Stop");
    }

    private GridBagConstraints gbcClone(GridBagConstraints gbc) {
        return gbcClone(gbc, null);
    }
    private GridBagConstraints gbcClone(GridBagConstraints gbc, Consumer<GridBagConstraints> edit) {
        GridBagConstraints clone = (GridBagConstraints) gbc.clone();
        if(edit != null){
            edit.accept(clone);
        }
        return clone;
    }

}

File: src/main/java/dev/cele/asa_sm/ui/components/ServerTab.kt
package dev.cele.asa_sm.ui.components

import dev.cele.asa_sm.components.ServerHelperComponent
import dev.cele.asa_sm.dto.AsaServerConfigDto
import dev.cele.asa_sm.ui.addValueChangedListener
import dev.cele.asa_sm.ui.frames.ProcessFrame
import dev.cele.asa_sm.ui.server_tab_accordions.AdministrationAccordion
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.awt.BorderLayout
import java.awt.GridBagConstraints
import java.awt.GridBagLayout
import java.awt.Label
import java.io.File
import java.nio.file.Path
import javax.swing.*


class ServerTab(val configDto: AsaServerConfigDto) : JPanel() {

    var log: Logger = LoggerFactory.getLogger(ServerTab::class.java)

    var installedVersionLabel = JLabel("-");
    var outDatedModsLabel = JLabel("-");
    var installationLocationLabel = JLabel("-");
    var serverStatusLabel = JLabel("-");
    var availabilityLabel = JLabel("-");
    var playerCount = JLabel("-/-");

    var scrollPaneContent = JPanel();

    val springHelper: ServerHelperComponent
        get() = ServerHelperComponent.get()

    init {
        //region initial setup, var scrollPaneContent = JPanel()
        layout = BorderLayout()

        val globalVerticalGBC = GridBagConstraints().apply {
            fill = GridBagConstraints.HORIZONTAL
            anchor = GridBagConstraints.BASELINE
            gridwidth = GridBagConstraints.REMAINDER
            weightx = 1.0
        }

        //create a JScrollPane and a content panel
        scrollPaneContent.apply {
            layout = GridBagLayout()
        }
        JScrollPane().apply {
            verticalScrollBarPolicy = JScrollPane.VERTICAL_SCROLLBAR_ALWAYS
            horizontalScrollBarPolicy = JScrollPane.HORIZONTAL_SCROLLBAR_NEVER
            viewport.add(scrollPaneContent)
        }.also {
            add(it, BorderLayout.CENTER)
        }
        //endregion

        GroupedPanel("").apply {
            layout = GridBagLayout()

            //region row 1: profile id and find button
            val gbcrow1 = GridBagConstraints().apply {
                fill = GridBagConstraints.BOTH
                anchor = GridBagConstraints.LINE_START
                gridy = 0
            }

            JLabel("Profile ID:").also {
                add(it, gbcrow1)
            }

            JLabel(configDto.guid).also {
                add(it, (gbcrow1.clone() as GridBagConstraints).apply { gridwidth = 5; weightx = 1.0 })
            }

            //find button
            JButton("Find").apply {
                addActionListener {
                    //find all the labels in the scrollPaneContent and all the childs
                    var labels = findAllLabels(scrollPaneContent)
                }
            }.also {
                add(it, gbcrow1)
            }

            //import/export button
            JButton("Import/Export").apply {
                addActionListener {

                }
            }.also {
                add(it, gbcrow1)
            }
            //endregion

            //region row 2: profile name and save button

            var gbcrow2 = (gbcrow1.clone() as GridBagConstraints).apply {
                gridy = 1
            }

            JLabel("Profile Name:").also {
                add(it, gbcrow2)
            }

            JTextField(configDto.profileName).apply {
                //add event listener for value change
                document.addValueChangedListener {
                    configDto.profileName = it
                    //find the jtabbedpane and change the title
                    val tabbedPane =  SwingUtilities.getAncestorOfClass(
                        JTabbedPane::class.java, this@ServerTab
                    ) as JTabbedPane

                    tabbedPane.setTitleAt(tabbedPane.selectedIndex, it)
                }
            }.also {
                add(it, (gbcrow2.clone() as GridBagConstraints).apply { gridwidth =6; weightx = 1.0 })
            }

            //save button
            JButton("Save").apply {
                addActionListener {
                    saveJson()
                }
            }.also {
                add(it, gbcrow2)
            }

            //endregion

            //region row 3:
            // left side: Installed version label, patch note button
            // center: outdated mods label, update mods button
            // right side: open server log folder button, verify/update server button

            var gbcrow3 = (gbcrow1.clone() as GridBagConstraints).apply {
                gridy = 2
            }

            //left side
            JLabel("Installed Version:").also {
                add(it, gbcrow3)
            }
            installedVersionLabel.also {
                add(it, gbcrow3)
            }

            JButton("Patch Notes").apply {
                addActionListener {

                }
            }.also {
                add(it, gbcrow3)
            }

            //center
            JLabel("Outdated Mods:").also {
                add(it, (gbcrow3.clone() as GridBagConstraints).apply { weightx = 1.0;})
            }
            outDatedModsLabel.also {
                add(it)
            }
            JButton("Update Mods").apply {
                addActionListener {

                }
            }.also {
                add(it, gbcrow3)
            }

            //right side
            JButton("Open Server Log Folder").apply {
                addActionListener {

                }
            }.also {
                add(it, gbcrow3)
            }
            JButton("Verify/Update Server").apply {
                addActionListener {
                    ProcessFrame.run(SwingUtilities.getWindowAncestor(this) as JFrame, springHelper.steamCMDService.downloadVerifyServerCommand("servers"+File.separator+configDto.guid));
                }
            }.also {
                add(it, gbcrow3)
            }

            //endregion


            //region row 4:
            //left side: Installation Location label
            //right side: open installation location button, set installation location button

            var gbcrow4 = (gbcrow1.clone() as GridBagConstraints).apply {
                gridy = 3
            }

            //left side
            JLabel("Installation Location: ").also {
                add(it, gbcrow4)
            }

            JLabel("servers" + File.separator + configDto.guid.toString()).also {
                add(it, (gbcrow4.clone() as GridBagConstraints).apply { gridwidth=6; weightx = 1.0 })
            }

            //right side
            JButton("Open Installation Location").apply {
                addActionListener {

                }
            }.also {
                add(it, gbcrow4)
            }

            //endregion

            //region row 5:
            //left side: server status label (double label)

            var gbcrow5 = (gbcrow1.clone() as GridBagConstraints).apply {
                gridy = 4
            }

            JLabel("Server Status:").also {
                add(it, gbcrow5)
            }
            serverStatusLabel.also {
                add(it, gbcrow5)
            }

            //center: availability label (double label)
            JLabel("Availability:").also {
                add(it, (gbcrow5.clone() as GridBagConstraints).apply { weightx = 1.0; })
            }
            availabilityLabel.also {
                add(it, gbcrow5)
            }

            //right side: player count label (double label), start button, rcon button
            JLabel("Player Count:").also {
                add(it, (gbcrow5.clone() as GridBagConstraints).apply { weightx = 1.0; })
            }
            playerCount.also {
                add(it, gbcrow5)
            }

            JButton("Start").apply {
                addActionListener {
                    springHelper.commandRunnerService.runCommand(*configDto.command)
                }
            }.also {
                add(it, gbcrow5)
            }
            JButton("RCON").apply {
                addActionListener {

                }
            }.also {
                add(it, gbcrow5)
            }


        } .also { scrollPaneContent.add(it, globalVerticalGBC)  }


        //create a group named "Administration"
        var administrationAccordion = AdministrationAccordion(configDto).apply {
            expanded = true
        }.also {
            scrollPaneContent.add(it, globalVerticalGBC)
        }

        //create a group named "Automatic Management"

        //create a group named "Server Details"

        //create a group named "Rules"

        //create a group named "Chat and Notifications"

        //create a group named "HUD and Visuals"

        //create a group named "Player Settings"

        //create a group named "Dino Settings"

        //create a group named "Environment"

        //create a group named "Structures"

        //create a group named "Engrams"

        //create a group named "Server File Details"

        //create a group named "Custom GameUserSettings.ini Settings"

        //create a group named "Custom Game.ini Settings"

        //create a group named "Player and Dino Level Progressions"

        //create a group named "Crafting Overrides"

        //create a group named "Stack Size Overrides"

        //create a group named "Map Spawner Overrides"

        //create a group named "Supply Crate Overrides"

        //create a group named "Exclude ItemIDs From Supply Crates Overrides"

        //create a group named "Prevent Transfer Overrides"


        //create an empty filler panel
        JPanel().apply {
            preferredSize = java.awt.Dimension(0, 0)
        }.also {
            scrollPaneContent.add(it, (globalVerticalGBC.clone() as GridBagConstraints).apply { weighty = 1.0 })
        }

    }

    private fun findAllLabels(pane: JPanel, labels: MutableList<Label> = mutableListOf()): MutableList<Label> {
        for (component in pane.components) {
            if (component is Label) {
                labels.add(component)
            } else if (component is JPanel) {
                findAllLabels(component, labels)
            }
        }
        return labels
    }

    private fun saveJson() {
        //save the configDto to a json file
        Path.of("data"+File.separator+"profiles"+File.separator+configDto.guid.toString()+".json")
            .toFile()
            .writeText(
                springHelper.objectMapper.writeValueAsString(configDto)
            )
        log.info("Saved configDto to"+File.separator+configDto.guid.toString()+".json file")
    }

}
